server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/StandaloneServerLocalConfigurationStorage.java

```diff
@@ -0,0 +1,142 @@
+      if (!flags.contains(CacheContainerAdmin.AdminFlag.PERMANENT)) {
+         throw new CacheConfigurationException("Cannot create a non-permanent cache");
+      }
+
+      if (template == null) {
+         throw new CacheConfigurationException("Cannot create a cache not based on an existing template");
+      }
+
+      String cacheMode = findCacheMode(template, true);
+
+      if (cacheMode == null) {
+         throw new CacheConfigurationException("No such template " + template);
+      }
+
+      final ModelNode cacheAddOp = new ModelNode();
+      cacheAddOp.get(OP).set(ADD);
+      cacheAddOp.get(OP_ADDR).set(rootPath.append(cacheMode, name).toModelNode());
+      cacheAddOp.get("configuration").set(template);
+
+      try {
+         ModelNode resp = modelControllerClient.execute(cacheAddOp);
+         if (!SUCCESS.equals(resp.get(OUTCOME).asString())) {
+            throw new CacheConfigurationException("Could not create cache");
+         }
+      } catch (IOException e) {
+         throw new CacheConfigurationException("Could not create cache", e);
+      }
+   }
+
+   @Override
+   public void removeCache(String name, EnumSet<CacheContainerAdmin.AdminFlag> flags) {
+      String cacheMode = findCacheMode(name, false);
+
+      if (cacheMode == null) {
+         throw new CacheConfigurationException("No such cache");
+      }
+
+      final ModelNode cacheRemoveOp = new ModelNode();
+      cacheRemoveOp.get(OP).set(REMOVE_OPERATION);
+      cacheRemoveOp.get(OP_ADDR).set(rootPath.append(cacheMode, name).toModelNode());
+
+      try {
+         ModelNode resp = modelControllerClient.execute(cacheRemoveOp);
+         if (!SUCCESS.equals(resp.get(OUTCOME).asString())) {
+            throw new CacheConfigurationException("Could not remove cache");
+         }
+      } catch (IOException e) {
+         throw new CacheConfigurationException("Could not remove cache", e);
+      }
+   }
+
+   private String findCacheMode(String name, boolean configuration) {
+      for (String cacheMode : CACHE_MODES) {
+         PathAddress address = rootPath;
+         if (configuration) {
+            address = address.append(CacheContainerConfigurationsResource.PATH).append(cacheMode + "-configuration", name);
+         } else {
+            address = address.append(cacheMode, name);
+         }
+         ModelNode op = new ModelNode();
+         op.get(OP).set(READ_ATTRIBUTE_OPERATION);
+         op.get(OP_ADDR).set(address.toModelNode());
+         op.get(NAME).set(configuration ? "template" : "configuration");
+         try {
+            ModelNode resp = modelControllerClient.execute(op);
+            if (SUCCESS.equals(resp.get(OUTCOME).asString())) {
+               return cacheMode;
+            }
+         } catch (IOException e) {
+            // Ignore and try the next
+         }
+      }
+      // Nothing found
+      return null;
+   }
```

Comment:
Do we prevent caches/templates of different types having the same name? If not then this method will always retrieve the first template encountered based upon the `CACHE_MODES` order.

Reply:
We should prevent it (that's the behaviour in embedded anyway), but this is not the PR to do it